### PR TITLE
[FLINK-17625] [table] Fix ArrayIndexOutOfBoundsException in AppendOnlyTopNFunction

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change
fix the bug of ArrayIndexOutOfBoundsException in AppendOnlyTopNFunction
## Brief change log
  - *fix the bug of ArrayIndexOutOfBoundsException in AppendOnlyTopNFunction*
  - *fix the bug of currentTopNum will always increase even remove retired element  in AppendOnlyTopNFunction#processElementWithoutRowNumber method*

## Verifying this change

This change is already covered by existing tests in TopNFunctionTestBase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: ( no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@flinkbot

**Comment by lsyldliu** [comment]
cc @wuchong

**Comment by lsyldliu** [comment]
cc @beyond1920

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java [hunk 1]
```diff
@@ -210,14 +210,19 @@ private void processElementWithoutRowNumber(RowData input, Collector<RowData> ou
 			RowData lastKey = lastEntry.getKey();
 			List<RowData> lastList = (List<RowData>) lastEntry.getValue();
 			// remove last one
-			RowData lastElement = lastList.remove(lastList.size() - 1);
-			if (lastList.isEmpty()) {
+			int size = lastList.size();
+			RowData lastElement = null;
+			if (size > 0) {
+				lastElement = lastList.get(size - 1);
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:216]:
```suggestion
				lastElement = lastList.remove(size - 1);
```

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:216]: Remove on the list use index will be more efficient than `buffer.remove`.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java [hunk 2]
```diff
@@ -210,14 +210,19 @@ private void processElementWithoutRowNumber(RowData input, Collector<RowData> ou
...
+			}
+			if (size <= 1) {
 				buffer.removeAll(lastKey);
 				dataState.remove(lastKey);
 			} else {
+				buffer.remove(lastKey, lastElement);
 				dataState.put(lastKey, lastList);
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:223]:
```suggestion
				dataState.put(lastKey, new ArrayList<>(lastList));
```

We should do a shallow copy for the `lastList`, see FLINK-17918.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java (hunk 1 above)
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:216]: Good point. I just find that maybe we can use `TopNBuffer#removeLast` directly here.

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:216]: Good idea, I don't want to break the atomicity of `TopNBuffer`,  so I don't want to add new method which can operate currentTopNum directly in this class,  prefer to use `TopNBuffer#removeLast`

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:216]: But, I find `TopNBuffer#removeLast` also iterate the list, the efficient problem also exist.

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java:216]: Yes. I think we should improve `TopNBuffer#removeLast`, we can call `remove(index)` and `get(index)` when it is a `List`.

**Comment by lsyldliu** [comment]
@wuchong Thanks for your review and I've addressed your comments.

**Comment by wuchong** [comment]
It looks good to me in general. I just push a minor commit to improve the code. Please have a look @lsyldliu.

**Comment by lsyldliu** [comment]
> minor

@wuchong , thanks for your work, it looks good to me.

**Pull request merged**